File: packages/core/src/idempotency.ts
/**
 * Idempotency-key support (the Stripe pattern).
 *
 * Lets clients safely retry mutating requests: the first request for a given
 * key is executed and its result cached; subsequent requests with the same key
 * return the cached result instead of re-running the side effect.
 *
 * The in-memory store below is for single-instance/dev use. In production back
 * this with a shared store (Redis) implementing the same interface.
 */

export interface StoredResponse {
  status: number;
  body: unknown;
}

export interface IdempotencyStore {
  get(key: string): Promise<StoredResponse | undefined>;
  set(key: string, value: StoredResponse, ttlMs: number): Promise<void>;
}

export class InMemoryIdempotencyStore implements IdempotencyStore {
  private readonly entries = new Map<string, { value: StoredResponse; expiresAt: number }>();

  async get(key: string): Promise<StoredResponse | undefined> {
    const entry = this.entries.get(key);
    if (!entry) return undefined;
    if (Date.now() > entry.expiresAt) {
      this.entries.delete(key);
      return undefined;
    }
    return entry.value;
  }

  async set(key: string, value: StoredResponse, ttlMs: number): Promise<void> {
    this.entries.set(key, { value, expiresAt: Date.now() + ttlMs });
  }
}

export interface IdempotencyOptions {
  ttlMs?: number;
}

/**
 * Run `operation` at most once per `key`. Returns the cached response on
 * repeat calls. If `key` is undefined the operation always runs (no caching).
 */
export async function withIdempotency(
  store: IdempotencyStore,
  key: string | undefined,
  operation: () => Promise<StoredResponse>,
  options: IdempotencyOptions = {},
): Promise<StoredResponse> {
  if (!key) return operation();

  const cached = await store.get(key);
  if (cached) return cached;

  const result = await operation();
  await store.set(key, result, options.ttlMs ?? 24 * 60 * 60 * 1000);
  return result;
}

File: python/getexp-core/getexp_core/idempotency.py
"""Idempotency-key support (the Stripe pattern).

Lets clients safely retry mutating requests. The in-memory store is for
single-instance/dev use; back it with Redis in production via the same Protocol.
"""

from __future__ import annotations

import time
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any, Protocol


@dataclass(frozen=True, slots=True)
class StoredResponse:
    status: int
    body: Any


class IdempotencyStore(Protocol):
    async def get(self, key: str) -> StoredResponse | None: ...
    async def set(self, key: str, value: StoredResponse, ttl_s: float) -> None: ...


class InMemoryIdempotencyStore:
    def __init__(self) -> None:
        self._entries: dict[str, tuple[StoredResponse, float]] = {}

    async def get(self, key: str) -> StoredResponse | None:
        entry = self._entries.get(key)
        if entry is None:
            return None
        value, expires_at = entry
        if time.monotonic() > expires_at:
            del self._entries[key]
            return None
        return value

    async def set(self, key: str, value: StoredResponse, ttl_s: float) -> None:
        self._entries[key] = (value, time.monotonic() + ttl_s)


async def with_idempotency(
    store: IdempotencyStore,
    key: str | None,
    operation: Callable[[], Awaitable[StoredResponse]],
    *,
    ttl_s: float = 24 * 60 * 60,
) -> StoredResponse:
    """Run ``operation`` at most once per ``key``; return the cached response on repeats."""
    if not key:
        return await operation()

    cached = await store.get(key)
    if cached is not None:
        return cached

    result = await operation()
    await store.set(key, result, ttl_s)
    return result
